File: app-controller/core/llama_cpp_manager.py
"""
llama_cpp 模型管理模块
- 管理 llama-cpp-python OpenAI 兼容服务器进程（启动/停止/切换）
- 支持多模型并行运行在不同端口
- 提供模型信息（显存需求、路径、支持的功能等）
- 模型切换后自动自测验证
"""

import os
import subprocess
import signal
import asyncio
import httpx
import logging
from typing import Dict, List, Optional, Any
from datetime import datetime
from core.cache_service import cache_service

logger = logging.getLogger("ai_controller.llama_cpp")

model_switch_lock = asyncio.Lock()

LLAMA_CPP_MODELS_BASE_PATH = "/mnt/pve_models"

LLAMA_CPP_SERVER_BIN = "python"

LLAMA_CPP_SERVER_MODULE = "llama_cpp.server"


class LlamaCppProcessManager:
    def __init__(self):
        self._processes: Dict[int, subprocess.Popen] = {}
        self._model_by_port: Dict[int, str] = {}
        self._port_by_model: Dict[str, int] = {}
        self._configs: Dict[str, Dict] = {}

    def register_model(self, model_name: str, config: Dict):
        self._configs[model_name] = config
        port = config.get("port")
        if port:
            self._port_by_model[model_name] = port
            self._model_by_port[port] = model_name

    def unregister_model(self, model_name: str):
        config = self._configs.pop(model_name, None)
        if config:
            port = config.get("port")
            if port:
                self._port_by_model.pop(model_name, None)
                self._model_by_port.pop(port, None)

    def get_model_port(self, model_name: str) -> Optional[int]:
        return self._port_by_model.get(model_name)

    def get_model_by_port(self, port: int) -> Optional[str]:
        return self._model_by_port.get(port)

    def start_server(self, model_name: str) -> bool:
        config = self._configs.get(model_name)
        if not config:
            logger.error(f"No config registered for model: {model_name}")
            return False

        port = config.get("port")
        model_path = config.get("model_path", os.path.join(LLAMA_CPP_MODELS_BASE_PATH, model_name))

        if port in self._processes:
            proc = self._processes[port]
            if proc.poll() is None:
                logger.info(f"llama_cpp server already running on port {port} for model {model_name}")
                return True
            else:
                self._cleanup_process(port)

        n_gpu_layers = config.get("n_gpu_layers", -1)
        ctx_size = config.get("ctx_size", 4096)
        n_threads = config.get("n_threads", os.cpu_count() or 4)
        host = config.get("host", "0.0.0.0")

        cmd = [
            LLAMA_CPP_SERVER_BIN, "-m", LLAMA_CPP_SERVER_MODULE,
            "--model", model_path,
            "--host", host,
            "--port", str(port),
            "--n_gpu_layers", str(n_gpu_layers),
            "--ctx_size", str(ctx_size),
            "--n_threads", str(n_threads),
        ]

        extra_args = config.get("extra_args", [])
        cmd.extend(extra_args)

        logger.info(f"Starting llama_cpp server for {model_name}: {' '.join(cmd)}")

        try:
            proc = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                preexec_fn=os.setpgrp if os.name != 'nt' else None,
            )
            self._processes[port] = proc
            logger.info(f"llama_cpp server started for {model_name} on port {port}, PID={proc.pid}")
            return True
        except Exception as e:
            logger.error(f"Failed to start llama_cpp server for {model_name}: {e}")
            return False

    def stop_server(self, model_name: str) -> bool:
        port = self._port_by_model.get(model_name)
        if not port:
            logger.warning(f"No port registered for model: {model_name}")
            return False

        proc = self._processes.get(port)
        if not proc:
            logger.info(f"llama_cpp server not running for model {model_name}")
            return True

        if proc.poll() is not None:
            self._cleanup_process(port)
            logger.info(f"llama_cpp server already stopped for model {model_name}")
            return True

        try:
            proc.terminate()
            try:
                proc.wait(timeout=10)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=5)

            self._cleanup_process(port)
            logger.info(f"llama_cpp server stopped for {model_name} on port {port}")
            return True
        except Exception as e:
            logger.error(f"Failed to stop llama_cpp server for {model_name}: {e}")
            return False

    def is_server_running(self, model_name: str) -> bool:
        port = self._port_by_model.get(model_name)
        if not port:
            return False

        proc = self._processes.get(port)
        if not proc:
            return False

        if proc.poll() is not None:
            self._cleanup_process(port)
            return False

        return True

    def get_server_status(self, model_name: str) -> Dict[str, Any]:
        port = self._port_by_model.get(model_name)
        running = self.is_server_running(model_name)
        proc = self._processes.get(port) if port else None

        return {
            "model": model_name,
            "port": port,
            "running": running,
            "pid": proc.pid if proc and running else None,
            "service": "llama_cpp",
        }

    def get_all_running_models(self) -> List[str]:
        running = []
        for model_name in list(self._configs.keys()):
            if self.is_server_running(model_name):
                running.append(model_name)
        return running

    def _cleanup_process(self, port: int):
        self._processes.pop(port, None)

    def cleanup_all(self):
        for port, proc in list(self._processes.items()):
            if proc.poll() is None:
                try:
                    proc.terminate()
                    proc.wait(timeout=5)
                except Exception:
                    proc.kill()
            self._cleanup_process(port)


async def test_llama_cpp_model(model_name: str, port: int, max_retries: int = 5, retry_delay: int = 5) -> Dict[str, Any]:
    url = f"http://localhost:{port}/v1/chat/completions"

    payload = {
        "model": model_name,
        "messages": [{"role": "user", "content": "Hello, please respond briefly."}],
        "max_tokens": 10,
        "temperature": 0.7,
    }

    for attempt in range(max_retries):
        try:
            async with httpx.AsyncClient(timeout=30) as client:
                response = await client.post(url, json=payload, timeout=30)

                if response.status_code == 200:
                    result = response.json()
                    if result.get("choices") and len(result["choices"]) > 0:
                        content = result["choices"][0].get("message", {}).get("content", "")
                        if content.strip():
                            return {
                                "success": True,
                                "message": "Model test passed",
                                "attempt": attempt + 1,
                                "response_content": content.strip(),
                            }
                        else:
                            return {
                                "success": False,
                                "message": "Model returned empty response",
                                "attempt": attempt + 1,
                            }
                    else:
                        return {
                            "success": False,
                            "message": "Invalid response structure",
                            "attempt": attempt + 1,
                        }
                else:
                    if attempt < max_retries - 1:
                        await asyncio.sleep(retry_delay)
                        continue
                    return {
                        "success": False,
                        "message": f"HTTP error {response.status_code}",
                        "attempt": attempt + 1,
                    }
        except Exception as e:
            if attempt < max_retries - 1:
                await asyncio.sleep(retry_delay)
                continue
            return {
                "success": False,
                "message": "Connection failed",
                "attempt": attempt + 1,
                "error": str(e),
            }

    return {
        "success": False,
        "message": "All retry attempts failed",
        "attempt": max_retries,
    }


def scan_gguf_models(base_path: str = LLAMA_CPP_MODELS_BASE_PATH) -> List[Dict[str, Any]]:
    cached = cache_service.get("ai_controller:cache:gguf_model_list")
    if cached is not None:
        return cached

    models = []
    if not os.path.exists(base_path):
        return models

    for model_name in sorted(os.listdir(base_path)):
        model_path = os.path.join(base_path, model_name)
        if not os.path.isdir(model_path):
            continue

        gguf_files = [
            f for f in os.listdir(model_path)
            if f.endswith(".gguf")
        ]

        if not gguf_files:
            continue

        primary_gguf = gguf_files[0]
        full_path = os.path.join(model_path, primary_gguf)

        model_info = {
            "name": model_name,
            "path": model_path,
            "gguf_file": primary_gguf,
            "gguf_path": full_path,
            "size_mb": _get_file_size_mb(full_path),
            "service": "llama_cpp",
        }

        models.append(model_info)

    cache_service.set("ai_controller:cache:gguf_model_list", models, ttl_seconds=60)
    return models


def _get_file_size_mb(filepath: str) -> int:
    try:
        return int(os.path.getsize(filepath) / (1024 * 1024))
    except Exception:
        return 0


llama_cpp_manager = LlamaCppProcessManager()

File: go-vllm-api/internal/service/llama_cpp_manager.go
package service

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
	"time"

	"go-vllm-api/internal/config"

	"go.uber.org/zap"
)

type LlamaCppServerConfig struct {
	Port        int      `json:"port" yaml:"port"`
	ModelPath   string   `json:"model_path" yaml:"model_path"`
	NGPULayers  int      `json:"n_gpu_layers" yaml:"n_gpu_layers"`
	CtxSize     int      `json:"ctx_size" yaml:"ctx_size"`
	NThreads    int      `json:"n_threads" yaml:"n_threads"`
	Host        string   `json:"host" yaml:"host"`
	ExtraArgs   []string `json:"extra_args" yaml:"extra_args"`
}

type LlamaCppProcessInfo struct {
	ModelName string            `json:"model_name"`
	Port      int               `json:"port"`
	PID       int               `json:"pid"`
	Running   bool              `json:"running"`
	StartedAt time.Time         `json:"started_at"`
	Service   string            `json:"service"`
	Config    LlamaCppServerConfig `json:"config"`
}

type GGUFModelInfo struct {
	Name      string `json:"name"`
	Path      string `json:"path"`
	GGUFFile  string `json:"gguf_file,omitempty"`
	SizeBytes int64  `json:"size_bytes"`
	SizeStr   string `json:"size"`
	SizeMB    int    `json:"size_mb"`
}

type LlamaCppManager struct {
	logger           *zap.Logger
	processes        map[string]*llamaCppProcess
	modelConfigs     map[string]*LlamaCppServerConfig
	portByModel      map[string]int
	modelByPort      map[int]string
	processesByPort  map[int]*llamaCppProcess
	mu               sync.Mutex
	requestClient    *http.Client
	cfg              *config.AppConfig
}

type llamaCppProcess struct {
	cmd       *exec.Cmd
	config    LlamaCppServerConfig
	pid       int
	started   time.Time
	modelName string
}

func NewLlamaCppManager(logger *zap.Logger, cfg *config.AppConfig) *LlamaCppManager {
	return &LlamaCppManager{
		logger:          logger,
		processes:       make(map[string]*llamaCppProcess),
		modelConfigs:    make(map[string]*LlamaCppServerConfig),
		portByModel:     make(map[string]int),
		modelByPort:     make(map[int]string),
		processesByPort: make(map[int]*llamaCppProcess),
		requestClient: &http.Client{
			Timeout: 30 * time.Second,
			Transport: &http.Transport{
				MaxIdleConns:        10,
				MaxIdleConnsPerHost: 5,
				IdleConnTimeout:     30 * time.Second,
			},
		},
		cfg: cfg,
	}
}

func (lm *LlamaCppManager) RegisterModelsFromConfig(cfg *config.AppConfig) {
	lm.mu.Lock()
	defer lm.mu.Unlock()

	lm.cfg = cfg
	for name, mc := range cfg.Models {
		if mc.Service == "llama_cpp" {
			llamaCfg := &LlamaCppServerConfig{
				Port:       mc.Port,
				ModelPath:  mc.ModelPath,
				NGPULayers: mc.NGPULayers,
				CtxSize:    mc.CtxSize,
				NThreads:   mc.NThreads,
				Host:       mc.Host,
				ExtraArgs:  mc.ExtraArgs,
			}
			if llamaCfg.NGPULayers == 0 {
				llamaCfg.NGPULayers = cfg.LlamaCpp.DefaultNGPULayers
			}
			if llamaCfg.CtxSize == 0 {
				llamaCfg.CtxSize = cfg.LlamaCpp.DefaultCtxSize
			}
			if llamaCfg.Host == "" {
				llamaCfg.Host = cfg.LlamaCpp.DefaultHost
			}
			if llamaCfg.ModelPath == "" {
				llamaCfg.ModelPath = filepath.Join(cfg.LlamaCpp.ModelsBasePath, name)
			}
			lm.modelConfigs[name] = llamaCfg
			lm.portByModel[name] = mc.Port
			lm.modelByPort[mc.Port] = name
		}
	}
	lm.logger.Info("registered llama_cpp models from config", zap.Int("count", len(lm.modelConfigs)))
}

func (lm *LlamaCppManager) RegisterModel(modelName string, cfg LlamaCppServerConfig) {
	lm.mu.Lock()
	defer lm.mu.Unlock()
	lm.modelConfigs[modelName] = &cfg
	lm.portByModel[modelName] = cfg.Port
	lm.modelByPort[cfg.Port] = modelName
}

func (lm *LlamaCppManager) UnregisterModel(modelName string) {
	lm.mu.Lock()
	defer lm.mu.Unlock()
	if cfg, ok := lm.modelConfigs[modelName]; ok {
		port := cfg.Port
		lm.portByModel[modelName] = 0
		lm.modelByPort[port] = ""
		delete(lm.modelConfigs, modelName)
	}
}

func (lm *LlamaCppManager) GetModelPort(modelName string) int {
	lm.mu.Lock()
	defer lm.mu.Unlock()
	if port, ok := lm.portByModel[modelName]; ok {
		return port
	}
	return 0
}

func (lm *LlamaCppManager) GetModelByPort(port int) string {
	lm.mu.Lock()
	defer lm.mu.Unlock()
	if name, ok := lm.modelByPort[port]; ok {
		return name
	}
	return ""
}

func (lm *LlamaCppManager) StartServerByName(ctx context.Context, modelName string) error {
	lm.mu.Lock()
	serverCfg, ok := lm.modelConfigs[modelName]
	if !ok {
		lm.mu.Unlock()
		return fmt.Errorf("no config registered for model: %s", modelName)
	}

	if proc, exists := lm.processes[modelName]; exists && proc.cmd.ProcessState == nil {
		lm.mu.Unlock()
		return fmt.Errorf("server for %s already running on port %d", modelName, serverCfg.Port)
	}

	cfg := *serverCfg
	lm.mu.Unlock()

	return lm.StartServer(ctx, modelName, cfg)
}

func (lm *LlamaCppManager) StartServer(ctx context.Context, modelName string, cfg LlamaCppServerConfig) error {
	lm.mu.Lock()
	if proc, exists := lm.processes[modelName]; exists && proc.cmd.ProcessState == nil {
		lm.mu.Unlock()
		return fmt.Errorf("server for %s already running", modelName)
	}

	nThreads := cfg.NThreads
	if nThreads <= 0 {
		nThreads = runtime.NumCPU()
	}

	serverModule := lm.cfg.LlamaCpp.ServerModule
	if serverModule == "" {
		serverModule = "llama_cpp.server"
	}

	args := []string{
		"-m", serverModule,
		"--model", cfg.ModelPath,
		"--port", fmt.Sprintf("%d", cfg.Port),
		"--n_gpu_layers", fmt.Sprintf("%d", cfg.NGPULayers),
		"--ctx_size", fmt.Sprintf("%d", cfg.CtxSize),
		"--n_threads", fmt.Sprintf("%d", nThreads),
		"--host", cfg.Host,
	}
	args = append(args, cfg.ExtraArgs...)

	cmd := exec.CommandContext(ctx, "python", args...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Start(); err != nil {
		lm.mu.Unlock()
		return fmt.Errorf("start llama_cpp server: %w", err)
	}

	proc := &llamaCppProcess{
		cmd:       cmd,
		config:    cfg,
		pid:       cmd.Process.Pid,
		started:   time.Now(),
		modelName: modelName,
	}
	lm.processes[modelName] = proc
	lm.processesByPort[cfg.Port] = proc
	lm.mu.Unlock()

	lm.logger.Info("llama_cpp server started",
		zap.String("model", modelName),
		zap.Int("port", cfg.Port),
		zap.Int("pid", cmd.Process.Pid))

	go func() {
		err := cmd.Wait()
		lm.mu.Lock()
		delete(lm.processes, modelName)
		delete(lm.processesByPort, cfg.Port)
		lm.mu.Unlock()
		if err != nil && err.Error() != "signal: killed" {
			lm.logger.Warn("llama_cpp server exited",
				zap.String("model", modelName),
				zap.Error(err))
		}
	}()

	return nil
}

func (lm *LlamaCppManager) StopServer(modelName string) error {
	lm.mu.Lock()
	proc, exists := lm.processes[modelName]
	if !exists {
		lm.mu.Unlock()
		return fmt.Errorf("no running server for %s", modelName)
	}
	port := proc.config.Port
	lm.mu.Unlock()

	if proc.cmd.Process != nil {
		proc.cmd.Process.Signal(os.Interrupt)
		time.Sleep(2 * time.Second)

		done := make(chan error, 1)
		go func() {
			done <- proc.cmd.Wait()
		}()

		select {
		case <-done:
		case <-time.After(5 * time.Second):
			proc.cmd.Process.Kill()
		}
	}

	lm.mu.Lock()
	delete(lm.processes, modelName)
	delete(lm.processesByPort, port)
	lm.mu.Unlock()

	lm.logger.Info("llama_cpp server stopped", zap.String("model", modelName))
	return nil
}

func (lm *LlamaCppManager) IsServerRunning(modelName string) bool {
	lm.mu.Lock()
	defer lm.mu.Unlock()
	proc, exists := lm.processes[modelName]
	if !exists {
		return false
	}
	if proc.cmd.ProcessState != nil {
		return false
	}
	return true
}

func (lm *LlamaCppManager) GetServerStatus() map[string]LlamaCppProcessInfo {
	lm.mu.Lock()
	defer lm.mu.Unlock()

	status := make(map[string]LlamaCppProcessInfo)
	for name, proc := range lm.processes {
		running := proc.cmd.ProcessState == nil
		pid := 0
		if proc.cmd.Process != nil {
			pid = proc.cmd.Process.Pid
		}
		status[name] = LlamaCppProcessInfo{
			ModelName: name,
			Port:      proc.config.Port,
			PID:       pid,
			Running:   running,
			StartedAt: proc.started,
			Service:   "llama_cpp",
			Config:    proc.config,
		}
	}
	return status
}

func (lm *LlamaCppManager) GetModelStatus(modelName string) LlamaCppProcessInfo {
	lm.mu.Lock()
	defer lm.mu.Unlock()

	port := lm.portByModel[modelName]
	running := lm.IsServerRunning(modelName)
	proc := lm.processes[modelName]
	pid := 0
	if proc != nil && proc.cmd.Process != nil && running {
		pid = proc.cmd.Process.Pid
	}

	cfg := LlamaCppServerConfig{}
	if mc, ok := lm.modelConfigs[modelName]; ok {
		cfg = *mc
	}

	return LlamaCppProcessInfo{
		ModelName: modelName,
		Port:      port,
		PID:       pid,
		Running:   running,
		Service:   "llama_cpp",
		Config:    cfg,
	}
}

func (lm *LlamaCppManager) GetAllRunningModels() []string {
	lm.mu.Lock()
	defer lm.mu.Unlock()
	var running []string
	for name := range lm.modelConfigs {
		if lm.IsServerRunning(name) {
			running = append(running, name)
		}
	}
	return running
}

func (lm *LlamaCppManager) ScanGGUFModels() []GGUFModelInfo {
	var models []GGUFModelInfo
	basePath := lm.cfg.LlamaCpp.ModelsBasePath

	entries, err := os.ReadDir(basePath)
	if err != nil {
		lm.logger.Error("scan gguf models", zap.String("path", basePath), zap.Error(err))
		return models
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		dirPath := filepath.Join(basePath, entry.Name())
		filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return nil
			}
			if !info.IsDir() && strings.HasSuffix(strings.ToLower(info.Name()), ".gguf") {
				sizeStr := formatFileSize(info.Size())
				sizeMB := int(info.Size() / (1024 * 1024))
				models = append(models, GGUFModelInfo{
					Name:      entry.Name(),
					Path:      dirPath,
					GGUFFile:  info.Name(),
					SizeBytes: info.Size(),
					SizeStr:   sizeStr,
					SizeMB:    sizeMB,
				})
			}
			return nil
		})
	}
	return models
}

type LlamaCppTestResult struct {
	Success        bool   `json:"success"`
	Message        string `json:"message"`
	Attempt        int    `json:"attempt"`
	ResponseContent string `json:"response_content,omitempty"`
	Error          string `json:"error,omitempty"`
}

func (lm *LlamaCppManager) TestModel(ctx context.Context, port int, modelName string) *LlamaCppTestResult {
	maxRetries := 5
	for attempt := 1; attempt <= maxRetries; attempt++ {
		lm.logger.Info("testing llama_cpp model",
			zap.String("model", modelName),
			zap.Int("port", port),
			zap.Int("attempt", attempt))

		result := lm.sendTestRequest(ctx, port, modelName)
		if result.Success {
			lm.logger.Info("llama_cpp model test passed", zap.String("model", modelName))
			return result
		}

		lm.logger.Warn("llama_cpp model test attempt failed",
			zap.Int("attempt", attempt), zap.String("message", result.Message))
		if attempt < maxRetries {
			time.Sleep(5 * time.Second)
		}
	}
	return &LlamaCppTestResult{
		Success: false,
		Message: fmt.Sprintf("llama_cpp model at port %d failed after %d attempts", port, maxRetries),
		Attempt: maxRetries,
	}
}

func (lm *LlamaCppManager) sendTestRequest(ctx context.Context, port int, modelName string) *LlamaCppTestResult {
	url := fmt.Sprintf("http://localhost:%d/v1/chat/completions", port)
	payload := map[string]interface{}{
		"model":   modelName,
		"messages": []map[string]interface{}{
			{"role": "user", "content": "Hello, please respond briefly."},
		},
		"max_tokens":  10,
		"temperature": 0.7,
		"stream":      false,
	}

	data, err := json.Marshal(payload)
	if err != nil {
		return &LlamaCppTestResult{Success: false, Message: fmt.Sprintf("marshal payload: %v", err), Attempt: 1}
	}

	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(data))
	if err != nil {
		return &LlamaCppTestResult{Success: false, Message: fmt.Sprintf("create request: %v", err), Attempt: 1}
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := lm.requestClient.Do(req)
	if err != nil {
		return &LlamaCppTestResult{Success: false, Message: fmt.Sprintf("test request: %v", err), Attempt: 1, Error: err.Error()}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return &LlamaCppTestResult{
			Success: false,
			Message: fmt.Sprintf("HTTP error %d: %s", resp.StatusCode, string(body)),
			Attempt: 1,
		}
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return &LlamaCppTestResult{Success: false, Message: fmt.Sprintf("decode response: %v", err), Attempt: 1}
	}

	choices, ok := result["choices"].([]interface{})
	if !ok || len(choices) == 0 {
		return &LlamaCppTestResult{Success: false, Message: "invalid response structure: no choices", Attempt: 1}
	}

	choice, ok := choices[0].(map[string]interface{})
	if !ok {
		return &LlamaCppTestResult{Success: false, Message: "invalid response structure: choice format", Attempt: 1}
	}

	msg, ok := choice["message"].(map[string]interface{})
	if !ok {
		return &LlamaCppTestResult{Success: false, Message: "invalid response structure: no message", Attempt: 1}
	}

	content, _ := msg["content"].(string)
	if content == "" {
		return &LlamaCppTestResult{Success: false, Message: "Model returned empty response", Attempt: 1}
	}

	return &LlamaCppTestResult{
		Success:         true,
		Message:         "Model test passed",
		ResponseContent: strings.TrimSpace(content),
	}
}

func (lm *LlamaCppManager) CleanupAll() {
	lm.mu.Lock()
	models := make([]string, 0, len(lm.processes))
	for name := range lm.processes {
		models = append(models, name)
	}
	lm.mu.Unlock()

	for _, name := range models {
		lm.StopServer(name)
	}
	lm.logger.Info("all llama_cpp servers cleaned up")
}

func formatFileSize(bytes int64) string {
	if bytes >= 1<<30 {
		return fmt.Sprintf("%.1f GB", float64(bytes)/float64(1<<30))
	}
	if bytes >= 1<<20 {
		return fmt.Sprintf("%.1f MB", float64(bytes)/float64(1<<20))
	}
	return fmt.Sprintf("%d KB", bytes/1024)
}
